instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/utils/Utils.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright The Hypertrace Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.utils;
+
+import io.undertow.server.HttpServerExchange;
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
+import org.hypertrace.agent.core.instrumentation.SpanAndObjectPair;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedBuffersFactory;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedByteArrayOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class Utils {
+
+  public static void handleTransfer(
+      final ByteBuffer byteBuffer, int numBytesRead, SpanAndObjectPair spanAndObjectPair) {
+    final Object associatedObject = spanAndObjectPair.getAssociatedObject();
+    final HttpServerExchange exchange;
+    if (associatedObject instanceof HttpServerExchange) {
+      exchange = (HttpServerExchange) associatedObject;
+    } else {
+      return;
+    }
+
+    if (!exchange.isRequestComplete()) {
+      return;
+    }
+    final BoundedByteArrayOutputStream boundedByteArrayOutputStream =
+        BoundedBuffersFactory.createStream(
+            byteBuffer.remaining(), Charset.forName(exchange.getRequestCharset()));
+    final byte[] dst = new byte[byteBuffer.remaining()];
+    byteBuffer.get(dst);
+    boundedByteArrayOutputStream.write(dst, 0, dst.length);
+    log.info("test");
+    final String body;
+    try {
+      body = boundedByteArrayOutputStream.toStringWithSuppliedCharset().trim();
```

Comment:
why trim? We do not call it anywhere else

Reply:
Yeah, this PR is definitely still a "draft" and this, in my opinion, is mostly what's holding it back. The `ByteBuffer` that is written to by the `org.xnio.channels.StreamSourceChannel.read(java.nio.ByteBuffer)` in my test is read twice. The first time, it reads in the 8 chars that make up the request body and puts them into the `java.nio.ByteBuffer`. The second time, no bytes are read (read returns -1). Right now, how the implementation works, is that we ignore reads until `exchange.isRequestComplete()` is true. Then, we read the bytes into from the ByteBuffer` into a `ByteArray`. However, we don't know how many bytes should actually be read in, since `numBytesRead` is `-1`. So, it reads the entire byte buffer, including null bytes, into the byte array and then writes that to the bounded stream, resulting in a reported request body like `echo=bar        ....` where ` ` is repeated for the entire remaining size of the `ByteBuffer`.

This is definitely not ideal. I'm just wrapping up a solution to instead capture the bytes only when they are written to the `ByteBuffer`. In this case, we create a read-only ByteBuffer and reset the position of the read-only ByteBuffer to where it was before the read, like this: 
```java
readOnlyBuffer.position(readOnlyBuffer.position() - numBytesRead); 
final byte[] dst = new byte[numBytesRead];
readOnlyBuffer.get(dst);
boundedByteArrayOutputStream.write(dst, 0, dst.length);
```

Then, we can remove this `trim` and save on the space we allocate in the byte array. In my example test, `numBytesRead` is `8` whereas buffer.getRemaining()` is ~16,000.

What are your thoughts on this solution? is there a better way to capture the contents of the ByteBuffer without modifying the ByteBuffer instance used by undertow? 

I'm wondering if we'll need to modify my existing implementation to put a `ByteBufferSpanPair` in the instrumentation context (like the servlet implementation) when `getRequestChannel` is invoked, using the `StreamSourceChannel.read` as the key Then, we can look up the that  `ByteBufferSpanPair` when StreamSourceChannel.read is invoked. Finally, we can add a `ExchangeCompletionListener` to the `HttpExchange` to convert the `BoundedByteArrayOutputStream` object to a `String` and set the `HTTP_REQUEST_BODY` attribute on the associated span. This would harden our implementation of request body capturing in the case where `StreamSourceChannel.read` is invoked multiple times. I'm not quite sure yet if this is a reasonable fear 🤔

Reply:
> This is definitely not ideal. I'm just wrapping up a solution to instead capture the bytes only when they are written to the ByteBuffer. In this case, we create a read-only ByteBuffer and reset the position of the read-only ByteBuffer to where it was before the read, like this:

All other instrumentations hook directly into buffer write methods - e.g. see java stream instrumentation. Definitely avoid wrapping of buffers :)

Reply:
Ah okay! Just to make sure I understand, you think we should instrument `java.nio.ByteBuffer.put` APIs and to write the provided `bytes`/`byte[]` to the `BoundedByteArrayOutputStream` if we find that the `ByteBuffer` instance was associated with a request body? Meaning we would: 

1. put the the `SpanAndBuffer` in the `InstrumentationContext` with the `StreamSourceChannel` instance as the key when `HttpServerExchange.getRequestChannel()` is invoked
1. put the the `SpanAndBuffer` in the `InstrumentationContext` with the `ByteBuffer` instance as the key when `StreamSourceChannel.read(ByteBuffer)` is invoked on a `StreamSourceChannel` in the `InstrumentationContext`
1.  Write to the `BoundedByteArrayOutputStream` when the `ByteBuffer.put` APIs are called on a `ByteBuffer` instance in the `InstrumentationContext`

Reply:
yes something like this, again the idea is to minimize the intrusiveness of our code. Also note that some frameworks use buffers that can be read multiple times or the buffer provides interceptors API, in this case it's preferrable to use these APIs.

Reply:
Just ran into something interesting with this approach! In the case of this use case on Undertow, we see this sequence of events the first time the request body is read:

### Approach 1: instrument `ByteBuffer` APIs, watching for `.put(....)` invocations on the `ByteBuffer` passed to `StreamSourceChannel.read(ByteBuffer)`.=
1. `HttpServerExchange.getRequestChannel()` is invoked. Our instrumentation properly puts the `StreamSourceChannel` in the `ContextStore<StreamSourceChannel, SpanAndObjectPair>` with the span
2. `StreamSourceChannel.read(ByteBuffer)` is called on the `StreamSourceChannel` instance returned from `HttpServerExchange.getRequestChannel()`. Now, we can put the `ByteBuffer` in a `ContextStore<ByteBuffer, SpanAndObjectPair>`
3. In the scope of `StreamSourceChannel.read(ByteBuffer)`, undertow delegates to a chain of object in order to properly read the request body from the `StreamSourceChannel` into the provided `ByteBuffer`. In the use-case of my test, this chain looks like [HttpServerExchange$ReadDispatchChannel.read(ByteBuffer)](https://github.com/undertow-io/undertow/blob/master/core/src/main/java/io/undertow/server/HttpServerExchange.java#L2419) -> [DetachableStreamSourceChannel.read(ByteBuffer)](https://github.com/undertow-io/undertow/blob/77c3b9bec4895f48aee3155747cefb9dccd2ec6b/core/src/main/java/io/undertow/channels/DetachableStreamSourceChannel.java#L209) -> [ConduitStreamSourceChannel.read(ByteBuffer)](https://github.com/xnio/xnio/blob/3.x/api/src/main/java/org/xnio/conduits/ConduitStreamSourceChannel.java#L127) -> [FixedLengthStreamSourceConduit.read(ByteBuffer)](https://github.com/undertow-io/undertow/blob/master/core/src/main/java/io/undertow/conduits/FixedLengthStreamSourceConduit.java#L244) -> [ReadDataStreamSourceConduit.read(ByteBuffer)](https://github.com/undertow-io/undertow/blob/master/core/src/main/java/io/undertow/conduits/ReadDataStreamSourceConduit.java#L60).
4. `ReadDataStreamSourceConduit.read(ByteBuffer)` gets the bytes associated with the request body in the form of a `PooledByteBuffer` from the `AbstractServerConnection` API `getExtraBytes()`. This `PooledByteBuffer` composes the src `ByteBuffer` and uses `ByteBuffer.put(ByteBuffer)` API to write to the buffer we've instrumented
5. `java.nio.ByteBuffer.put(ByteBuffer)` delegates to `java.nio.ByteBuffer.put(byte)` for each `byte` in the src `ByteBuffer`. Unfortunately, not all `ByteBuffer`s use this implementation. Notably, the `ByteBuffer` instance here is a `java.nio.DirectByteBuffer`, which overrides the `put(ByteBuffer)` API and used `Unsafe` to directly copy the memory from the source `ByteBuffer` to the destination `ByteBuffer`, so instrumentation of the other `ByteBuffer.put(...)` APIs will not help us capture the request body 


### Approach 2: Instrument `ByteBuffer`, watching for `.put(....)` invocations on the `ByteBuffer` used by the Undertow `HttpReadListener`

1. This led me down the other path, capture the request body when it is being read into a `ByteBuffer` the very first time by undertow in [HttpReadListener](https://github.com/undertow-io/undertow/blob/master/core/src/main/java/io/undertow/server/protocol/http/HttpReadListener.java#L158). 
2. The above line delegates to the [ConduitStreamSourceChannel.read(ByteBuffer)](https://github.com/xnio/xnio/blob/3.x/api/src/main/java/org/xnio/conduits/ConduitStreamSourceChannel.java#L127) in xnio followed by [NioSocketConduit.read(ByteBuffer)](https://github.com/xnio/xnio/blob/3.x/nio-impl/src/main/java/org/xnio/nio/NioSocketConduit.java#L286). 
3. `NioSocketConduit` then delegates to the `java.nio.channels.SocketChannel` implementation [sun.nio.ch.SocketChannelImpl.read(ByteBuffer)](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java#L417). 
4. The above line uses a `NativeDispatcher` to natively read from the `SocketChannelImpl` into the destination `ByteBuffer`, again sidestepping the public `ByteBuffer` APIs.

Given those two situations, I don't think that watching for `.put(..)` invocations on a `ByteBuffer` will help us the same way it did for `InputStream`s. I think no matter what, we're going to have to look at the `ByteBuffer` instance after `read` has occurred and pull the request body information out of it based on how many bytes were read. Right now, the instrumentation we have applied in this PR captures the body in the context of the `HttpServerExchange.ReadDispatchChannel.read()` API, but I think we would do better to have it captured in the context of the read in `HttpReadListener.handleEvent` because undertow goes through significant lengths to make sure that it is only invoked once per request. Functionally though, the approach would be pretty similar to what we have now

Reply:
> All other instrumentations hook directly into buffer write methods - e.g. see java stream instrumentation. Definitely avoid wrapping of buffers :)

Also, circling back on this, note that the approach implemented here doesn't wrap the buffer in any way. We instrument the `org.xnio.channels.StreamSourceChannel.read(ByteBuffer)` API to capture the request body on method exit after it has been written to the `ByteBuffer`. In order to prevent modification of the destination `ByteBuffer`, before reading the data out from the `ByteBuffer` into our `BoundedByteArrayOutputStream`, I call `ByteBuffer.asReadOnlyBuffer` as a defensive measure to prevent modification of the original. 

```java

    @Advice.OnMethodExit(suppress = Throwable.class)
    public static void enter(
        @Advice.Enter int callDepth,
        @Advice.Return final int numBytesRead,
        @Advice.This final StreamSourceChannel streamSourceChannel,
        @Advice.Argument(0) final ByteBuffer byteBuffer) {
      if (callDepth > 0) {
        return;
      }
      HypertraceCallDepthThreadLocalMap.reset(StreamSourceChannel.class);
      final ContextStore<StreamSourceChannel, SpanAndBuffer> contextStore =
          InstrumentationContext.get(StreamSourceChannel.class, SpanAndBuffer.class);
      final SpanAndBuffer spanAndBuffer = contextStore.get(streamSourceChannel);
      if (spanAndBuffer != null) {
        Utils.handleRead(byteBuffer.asReadOnlyBuffer(), numBytesRead, spanAndBuffer);
      }
    }
```

Note here that the original `ByteBuffer` is not modified, just that the instance we pass to `Utils.handleRead` for analysis is designated read-only to prevent modification